cpp/src/arrow/io/memory.h

```diff
@@ -38,11 +42,49 @@ class Status;
 
 namespace io {
 
+class ARROW_EXPORT Memcopy {
+public:
+  virtual void memcopy(uint8_t* dst, const uint8_t* src, uint64_t nbytes) = 0;
```

Comment:
This is missing a virtual destructor and I believe will leak memory. I don't think we need these classes, though, only some configurable members on the Writer class